### DISCUSSION TITLE
In API or SDK: The Array-Keys should be the primaryKey.

### DISCUSSION THREAD

**Discussion by FreeSynergy** [directus/directus#22363] [Feature Requests]
### Summary

Question:
Why is an array returned in SDK or API responses instead of using the PrimaryKey as the key? After all, primary keys exist for this purpose...

---

When receiving responses from SDKs or APIs, it's common to encounter arrays of objects. However, a question arises: why aren't these arrays transformed into objects where the PrimaryKey serves as the key?

This approach seems more intuitive, especially when dealing with datasets where each object has a unique identifier, such as a PrimaryKey. By using the PrimaryKey as the key in an object, accessing and manipulating data becomes more straightforward and efficient.

For instance, instead of iterating through an array to find a specific object by its PrimaryKey, accessing it directly through an object would provide faster and more direct access.

Using the PrimaryKey as the key in an object can enhance readability, simplify data manipulation, and improve performance, especially in scenarios where quick data retrieval is crucial.

This approach is particularly crucial in scenarios like translations, where each translation may have a unique identifier, making it essential to quickly access specific translations by their keys.

Therefore, it might be worth considering this approach in SDKs and APIs to streamline data handling and improve developer experience.

### Basic Example

Certainly, here's the revised example using the `code` attribute as the key instead of numerical IDs:

### Before:
```javascript
// Response from API
const translationsArray = [
    { code: 'en', text: 'Hello' },
    { code: 'fr', text: 'Bonjour' },
    { code: 'es', text: 'Hola' }
];

// Accessing translation by iterating through the array
const findTranslationByCode = (translations, code) => {
    return translations.find(translation => translation.code === code);
};

const translation = findTranslationByCode(translationsArray, 'fr');
console.log(translation); // Output: { code: 'fr', text: 'Bonjour' }
```

### After:
```javascript
// Response from API transformed into an object with code as key
const translationsObject = {
    'en': { text: 'Hello' },
    'fr': { text: 'Bonjour' },
    'es': { text: 'Hola' }
};

// Accessing translation directly through the object using the code as key
const translationByKey = translationsObject['fr'];
console.log(translationByKey); // Output: { text: 'Bonjour' }
```

In the "Before" example, translations are returned as an array. To find a translation by its code, a function iterates through the array, comparing each object's code until a match is found.

In the "After" example, translations are transformed into an object where the codes serve as keys. Accessing a translation by its code is simplified and more efficient, as it involves directly accessing the object using the code as the key. This approach is particularly advantageous when dealing with large datasets or frequent data retrieval operations.

### Motivation

We write this to propose an enhancement to the data structure in SDKs or APIs. The motivation behind it is to improve efficiency and user experience when handling the received data.

By using an associative array where the PrimaryKey serves as the key, developers can retrieve data more quickly and easily. This leads to improved performance and a clearer data structure, especially in scenarios like translations where fast and efficient data manipulation is crucial.

The aim is to enhance the developer experience and support the development of more efficient applications. This change would simplify the development process and reduce potential errors by providing a more intuitive and efficient method of data access.

### Detailed Design

Sure, let's delve into a more detailed design proposal:

### Current State:
- SDKs or APIs typically return data as an array of objects.
- Each object in the array contains various attributes, including a PrimaryKey (e.g., ID or code).

### Proposed Enhancement:
- Transform the array of objects into an associative array where the PrimaryKey serves as the key.
- Retain all attributes of each object within the associative array.

### Implementation Steps:
1. **Data Transformation:**
   - Iterate over the array of objects.
   - Use the PrimaryKey attribute of each object as the key in the associative array.
   - Store the entire object as the value associated with each key.

2. **Usage:**
   - Access data directly using the PrimaryKey as the key in the associative array.
   - Data retrieval will be more efficient, as it avoids iterating through the entire array to find a specific object.

3. **Example:**
   ```javascript
   // Response from API transformed into an associative array with code as key
   const translationsObject = {
       'en': { text: 'Hello' },
       'fr': { text: 'Bonjour' },
       'es': { text: 'Hola' }
   };

   // Accessing translation directly through the object using the code as key
   const translationByKey = translationsObject['fr'];
   console.log(translationByKey); // Output: { text: 'Bonjour' }
   ```

### Benefits:
- **Improved Efficiency:** Direct access to data using the PrimaryKey reduces lookup time, especially in large datasets.
- **Enhanced Readability:** Using the PrimaryKey as the key in the associative array makes data retrieval more intuitive and readable.
- **Simplified Data Handling:** Developers can manipulate data more easily without needing to iterate through arrays.

### Considerations:
- **Compatibility:** Ensure that existing code relying on array structures is not negatively impacted by the proposed change.
- **Documentation:** Update SDK/API documentation to reflect the new data structure for clarity and ease of use.

### Conclusion:
By implementing this enhancement, SDKs and APIs can offer a more efficient and user-friendly data retrieval mechanism. Developers will benefit from improved performance and simplified data handling, ultimately enhancing the overall developer experience.

### Requirements List

Certainly! Here's a requirement list for the proposed enhancement:

### Requirement List for Transforming SDK/API Response into an Associative Array:

1. **Data Transformation:**
   - [ ] Transform the array of objects returned by the SDK/API into an associative array.
   - [ ] Use the PrimaryKey attribute of each object as the key in the associative array.
   - [ ] Store the entire object as the value associated with each key.

2. **Usage:**
   - [ ] Enable direct access to data using the PrimaryKey as the key in the associative array.
   - [ ] Ensure that data retrieval is efficient, avoiding the need to iterate through the entire array.

3. **Example Implementation:**
   - [ ] Provide clear examples illustrating how to access data using the new associative array structure.
   - [ ] Include sample code snippets demonstrating the benefits of direct data access.

4. **Compatibility:**
   - [ ] Ensure compatibility with existing code that relies on array structures.
   - [ ] Implement the transformation in a way that does not negatively impact backward compatibility.

5. **Documentation:**
   - [ ] Update SDK/API documentation to reflect the new data structure.
   - [ ] Include detailed explanations and examples to guide developers in utilizing the associative array.

6. **Testing:**
   - [ ] Conduct thorough testing to verify the correctness and efficiency of the data transformation.
   - [ ] Test scenarios with various sizes of data sets to ensure scalability and performance.

7. **Feedback and Iteration:**
   - [ ] Gather feedback from developers using the SDK/API regarding the new data structure.
   - [ ] Iterate based on feedback to address any issues or concerns raised by users.

8. **Release and Deployment:**
   - [ ] Plan the release of the enhanced SDK/API with the new data structure.
   - [ ] Ensure proper deployment procedures are followed to minimize disruptions for existing users.

9. **Monitoring and Maintenance:**
   - [ ] Monitor the performance of the enhanced SDK/API in production environments.
   - [ ] Address any performance issues or bugs that may arise post-release.

10. **Communication:**
    - [ ] Communicate the benefits of the enhanced data structure to developers.
    - [ ] Provide resources and support to assist developers in transitioning to the new data handling mechanism.

By fulfilling these requirements, the transformation of SDK/API responses into associative arrays will be implemented effectively, providing developers with a more efficient and user-friendly data retrieval experience.

### Drawbacks

While the proposed enhancement offers several benefits, it's essential to consider potential drawbacks:

1. **Increased Memory Usage:**
   - Storing data in an associative array may consume more memory compared to arrays, especially for large datasets. Each object's key-value pair in the associative array adds overhead.

2. **Potential Data Redundancy:**
   - If multiple objects share the same PrimaryKey, storing them in an associative array could lead to data redundancy. This redundancy may increase memory usage and complicate data management.

3. **Limited Flexibility:**
   - Associative arrays are less flexible than arrays in terms of data ordering. They do not maintain the original order of elements, which may be important in certain scenarios.

4. **Compatibility Concerns:**
   - Existing code relying on array structures may need to be updated to accommodate the new data format. This could introduce compatibility issues and require additional development effort.

5. **Performance Impact:**
   - While direct access to data using the PrimaryKey may improve performance, the overhead of managing associative arrays could impact performance negatively, particularly in resource-constrained environments or for frequent data manipulation operations.

6. **Complexity in Nested Structures:**
   - If the SDK/API response includes nested structures or arrays within objects, transforming them into associative arrays may introduce complexity, requiring careful handling and potentially nested associative arrays.

7. **Maintenance Overhead:**
   - Maintaining the consistency and correctness of the data transformation process, especially across different SDK versions or API endpoints, may require additional effort and ongoing maintenance.

8. **Learning Curve for Developers:**
   - Developers accustomed to working with array structures may need time to adjust to the new data format and understand the implications of using associative arrays, potentially leading to a learning curve.

9. **Debugging Challenges:**
   - Debugging code that interacts with associative arrays may be more challenging compared to arrays, as developers need to ensure proper handling of keys and values.

10. **Documentation Updates:**
    - Updating SDK/API documentation to reflect the new data structure requires careful attention to detail. Inadequate or inaccurate documentation could lead to confusion and errors for developers.

While these drawbacks should be considered, they can often be mitigated with careful planning, thorough testing, and effective communication with developers using the SDK/API. Additionally, the potential benefits of improved data access and manipulation may outweigh these drawbacks in many cases.

### Alternatives

Certainly! Here are some alternatives to consider instead of or in addition to transforming SDK/API responses into associative arrays:

1. **Custom Data Structures:**
   - Implement custom data structures tailored to specific use cases. For example, if quick data retrieval by PrimaryKey is crucial, design a specialized data structure optimized for this purpose.

2. **Pagination and Filtering:**
   - Instead of transforming the entire dataset into an associative array, consider implementing pagination and filtering mechanisms in the SDK/API. This allows developers to request only the data they need, reducing memory usage and improving performance.

3. **Caching:**
   - Implement caching mechanisms to store frequently accessed data in memory. This can help reduce the need for repeated data retrieval operations and improve overall performance.

4. **Data Indexing:**
   - Use indexing techniques to optimize data retrieval by specific attributes. For example, maintain indexes for commonly queried attributes to facilitate efficient lookup operations.

5. **Lazy Loading:**
   - Adopt lazy loading strategies to load data on-demand rather than loading the entire dataset upfront. This can help minimize memory usage and improve response times, especially for large datasets.

6. **Hybrid Approaches:**
   - Combine array-based and associative array-based data structures as needed. Use associative arrays for frequently accessed data or data with unique identifiers, while retaining arrays for sequential data or data with ordering requirements.

7. **Client-Side Transformations:**
   - Allow developers to choose between different data formats based on their preferences and use cases. Provide utilities or helper functions in the SDK/API to facilitate client-side transformations as needed.

8. **API Versioning:**
   - Introduce changes gradually through API versioning. Offer both the existing array-based response format and the new associative array-based format in different API versions, allowing developers to transition at their own pace.

9. **Feedback and Collaboration:**
   - Solicit feedback from developers using the SDK/API to understand their requirements and preferences regarding data structures. Collaborate with the developer community to identify the most effective solutions.

10. **Documentation and Education:**
    - Provide comprehensive documentation, tutorials, and examples demonstrating different data handling approaches. Educate developers on best practices for data retrieval, manipulation, and optimization.

By exploring these alternatives, you can tailor the data handling mechanisms in SDKs and APIs to meet diverse use cases and developer preferences effectively. Each approach has its strengths and considerations, so it's essential to evaluate them based on specific requirements and constraints.

### Adoption Strategy

Implementing an adoption strategy is crucial to ensure that developers effectively transition to the new data handling mechanism in SDKs and APIs. Here's a step-by-step adoption strategy:

1. **Communicate the Benefits:**
   - Clearly communicate the benefits of the new data structure to developers. Highlight how it improves performance, simplifies data retrieval, and enhances the overall developer experience.

2. **Provide Documentation:**
   - Update SDK/API documentation to include detailed explanations of the new data structure, along with examples illustrating its usage. Ensure that developers have access to comprehensive resources to understand the changes.

3. **Offer Migration Guides:**
   - Create migration guides or tutorials to assist developers in transitioning from the old data structure to the new one. Provide step-by-step instructions, code examples, and best practices for updating existing code.

4. **Enable Feature Flags:**
   - Implement feature flags or toggles to enable developers to opt into using the new data structure gradually. This allows for controlled testing and adoption, minimizing disruption to existing workflows.

5. **Provide Developer Support:**
   - Offer developer support through forums, documentation updates, and direct assistance channels. Address any questions, concerns, or issues raised by developers during the adoption process promptly and effectively.

6. **Run Pilot Programs:**
   - Conduct pilot programs with a select group of developers or partners to gather feedback and validate the effectiveness of the new data structure in real-world scenarios. Incorporate insights from pilot testing to refine the adoption strategy.

7. **Training and Education:**
   - Provide training sessions or workshops to educate developers on the advantages and usage of the new data structure. Offer hands-on exercises and demonstrations to reinforce learning and build confidence in adopting the changes.

8. **Monitor Adoption Metrics:**
   - Track adoption metrics, such as the number of developers using the new data structure, feedback received, and performance improvements observed. Use these metrics to assess the effectiveness of the adoption strategy and identify areas for improvement.

9. **Iterate Based on Feedback:**
   - Continuously gather feedback from developers and iterate on the adoption strategy based on their input. Address any challenges or barriers to adoption proactively, and adjust the strategy as needed to ensure successful implementation.

10. **Celebrate Successes:**
    - Celebrate milestones and successes achieved during the adoption process. Recognize and reward developers who actively participate in the transition and contribute to its success, fostering a positive and supportive community.

By following this adoption strategy, SDKs and APIs can facilitate a smooth transition to the new data handling mechanism, ensuring that developers can leverage its benefits effectively while minimizing disruption to their workflows.

### Unresolved Questions

None

**Comment by br41nslug** [comment]
This idea only works if the ID is requesting by the user, which is entirely optional. Plus it will add a noticable overhead of conditionally and recursively post-processing the data returned from the database into a different if the ID is present.

I'll be closing this request since we would like the structure of the API to be consistent independent of whether or not the IDs are requested.

**Reply by FreeSynergy** [reply]
This idea remains effective regardless of whether the user requests the ID. You can either include a parameter to toggle this behavior or additionally provide the primary key as both a key and an element if desired.

Why would this entail overhead?
Instead of pushing to create a new array, automatically setting the PK as the key in the array would not significantly increase processing time.

On the contrary:
Every search operation to find an item by its primary key takes longer if all objects need to be searched because they are unaware of the presence of the primary key.
Let's be honest:
This is precisely why we have primary keys, and it's the speed advantage we gain. That's why we implement it. (At some point, you may also consider that a primary key can function across multiple columns/fields, which also has its advantages...)

(Based on your description referring to it as an ID rather than a primary key, it seems you may not fully grasp the advantage. If I'm mistaken, I apologize, but it appears to be the case.)

Please reconsider incorporating this feature without closing it.

Best regards,
Kal
